diff --git a/arch/arm64/boot/dts/freescale/mono-gateway-dk.dts b/arch/arm64/boot/dts/freescale/mono-gateway-dk.dts
index 0000000..0000000 100644
--- a/arch/arm64/boot/dts/freescale/mono-gateway-dk.dts
+++ b/arch/arm64/boot/dts/freescale/mono-gateway-dk.dts
@@ -847,6 +847,8 @@
 			dpaa-ethernet@0 {
 				compatible = "fsl,dpa-ethernet";
 				fsl,fman-mac = <&enet6>;
+				fsl,qman-frame-queues-rx = <0x40c 1 0x40d 1>;
+				fsl,qman-frame-queues-tx = <0x40e 1 0x40f 1>;
 				fsl,bman-buffer-pools = <&dpaa_pool>;
 				dma-coherent;
 			};
@@ -863,6 +865,8 @@
 			dpaa-ethernet@0 {
 				compatible = "fsl,dpa-ethernet";
 				fsl,fman-mac = <&enet7>;
+				fsl,qman-frame-queues-rx = <0x410 1 0x411 1>;
+				fsl,qman-frame-queues-tx = <0x412 1 0x413 1>;
 				fsl,bman-buffer-pools = <&dpaa_pool>;
 				dma-coherent;
 			};
