diff --git a/drivers/net/usb/usbnet.c b/drivers/net/usb/usbnet.c
--- a/drivers/net/usb/usbnet.c
+++ b/drivers/net/usb/usbnet.c
@@ -31,6 +31,15 @@
 #include <linux/kernel.h>
 #include <linux/pm_runtime.h>
 
+#ifdef CONFIG_CPE_FAST_PATH
+/*
+ * We alllocate extra 64 Bytes to reserve headroom in the sk_buff
+ * To be used by Fast Path (Head Room must be 4 Byte aligned
+ * because USB 2.0 controller doesn't support 2 byte alignment
+ */
+#define C2K_USBNET_SKB_HEADROOM_FAST_PATH 64
+#endif
+
 /*-------------------------------------------------------------------------*/
 
 /*
@@ -504,11 +513,19 @@ static int rx_submit (struct usbnet *dev
 		usb_free_urb(urb);
 		return -ENOLINK;
 	}
-
+#ifdef CONFIG_CPE_FAST_PATH
+	/* 
+	* We alllocate extra 64 Bytes to reserve headroom in the sk_buff 
+	* To be used by Fast Path (Head Room must be 4 Byte aligned 
+	* because USB 2.0 controller doesn't support 2 byte alignment
+	*/
+	skb = __netdev_alloc_skb(dev->net, size + C2K_USBNET_SKB_HEADROOM_FAST_PATH, GFP_ATOMIC);
+#else
 	if (test_bit(EVENT_NO_IP_ALIGN, &dev->flags))
 		skb = __netdev_alloc_skb(dev->net, size, flags);
 	else
 		skb = __netdev_alloc_skb_ip_align(dev->net, size, flags);
+#endif
 	if (!skb) {
 		netif_dbg(dev, rx_err, dev->net, "no rx skb\n");
 		usbnet_defer_kevent (dev, EVENT_RX_MEMORY);
@@ -516,6 +533,10 @@ static int rx_submit (struct usbnet *dev
 		return -ENOMEM;
 	}
 
+#ifdef CONFIG_CPE_FAST_PATH
+	skb_reserve (skb, C2K_USBNET_SKB_HEADROOM_FAST_PATH);
+#endif    
+
 	entry = (struct skb_data *) skb->cb;
 	entry->urb = urb;
 	entry->dev = dev;
diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -2417,6 +2417,10 @@ struct net_device {
 /* TCP minimal MSS is 8 (TCP_MIN_GSO_SIZE),
  * and shinfo->gso_segs is a 16bit field.
  */
+
+#if defined(CONFIG_CPE_FAST_PATH)
+	bool				wifi_offload_dev;
+#endif
 #define GSO_MAX_SIZE		(8 * GSO_MAX_SEGS)
 
 #define TSO_LEGACY_MAX_SIZE	65536
@@ -3365,10 +3369,31 @@ u16 dev_pick_tx_zero(struct net_device *
 int __dev_queue_xmit(struct sk_buff *skb, struct net_device *sb_dev);
 int __dev_direct_xmit(struct sk_buff *skb, u16 queue_id);
 
+#if defined(CONFIG_CPE_FAST_PATH)
+int original_dev_queue_xmit(struct sk_buff *skb);
+typedef int (*dpaa_wifi_xmit_local_hook_t)(struct sk_buff *skb);
+int dpa_register_wifi_xmit_local_hook(dpaa_wifi_xmit_local_hook_t hookfn);
+void dpa_unregister_wifi_xmit_local_hook(void);
+int dpa_add_dummy_eth_hdr(struct sk_buff** skb_in, int priv_headroom, unsigned char *hdroom_realloc);
+extern dpaa_wifi_xmit_local_hook_t dpaa_wifi_xmit_local_ipsec_handler;
+
+static inline int dev_queue_xmit(struct sk_buff *skb)
+{
+#if defined(CONFIG_INET_IPSEC_OFFLOAD) || defined(CONFIG_INET6_IPSEC_OFFLOAD)
+if (dpaa_wifi_xmit_local_ipsec_handler && skb->ipsec_offload &&
+    skb->dev->wifi_offload_dev)
+return dpaa_wifi_xmit_local_ipsec_handler(skb);
+#else
+(void)dpaa_wifi_xmit_local_ipsec_handler;
+#endif
+return original_dev_queue_xmit(skb);
+}
+#else
 static inline int dev_queue_xmit(struct sk_buff *skb)
 {
 	return __dev_queue_xmit(skb, NULL);
 }
+#endif
 
 static inline int dev_queue_xmit_accel(struct sk_buff *skb,
 				       struct net_device *sb_dev)
@@ -5647,6 +5672,14 @@ extern struct list_head ptype_base[PTYPE
 
 extern struct net_device *blackhole_netdev;
 
+#if defined(CONFIG_CPE_FAST_PATH)
+/* hook function for getting statistics from offloaded interfaces */
+typedef void (*fp_iface_stats_get)(struct net_device *device, struct rtnl_link_stats64 *tot);
+
+void dev_fp_stats_get_register(fp_iface_stats_get func);
+void dev_fp_stats_get_deregister(void);
+#endif
+
 /* Note: Avoid these macros in fast path, prefer per-cpu or per-queue counters. */
 #define DEV_STATS_INC(DEV, FIELD) atomic_long_inc(&(DEV)->stats.__##FIELD)
 #define DEV_STATS_ADD(DEV, FIELD, VAL) 	\
diff --git a/include/linux/poll.h b/include/linux/poll.h
--- a/include/linux/poll.h
+++ b/include/linux/poll.h
@@ -116,7 +116,6 @@ extern int poll_select_set_timeout(struc
 
 #define __MAP(v, from, to) \
 	(from < to ? (v & from) * (to/from) : (v & from) / (from/to))
-
 static inline __u16 mangle_poll(__poll_t val)
 {
 	__u16 v = (__force __u16)val;
diff --git a/include/net/ip.h b/include/net/ip.h
--- a/include/net/ip.h
+++ b/include/net/ip.h
@@ -737,6 +737,9 @@ enum ip_defrag_users {
 	IP_DEFRAG_VS_FWD,
 	IP_DEFRAG_AF_PACKET,
 	IP_DEFRAG_MACVLAN,
+#ifdef CONFIG_CPE_TNL_4RD
+	IP_DEFRAG_IP6_TNL_4RD /* Used to support Post Fragmentation for 4o6 tunnels */
+#endif
 };
 
 /* Return true if the value of 'user' is between 'lower_bond'
diff --git a/include/net/ip6_tunnel.h b/include/net/ip6_tunnel.h
--- a/include/net/ip6_tunnel.h
+++ b/include/net/ip6_tunnel.h
@@ -52,7 +52,12 @@ struct ip6_tnl {
 	struct flowi fl;	/* flowi template for xmit */
 	struct dst_cache dst_cache;	/* cached dst */
 	struct gro_cells gro_cells;
-
+#if defined(CONFIG_INET6_IPSEC_OFFLOAD)
+	u32 genid;
+#endif
+#ifdef CONFIG_CPE_4RD_TUNNEL
+	struct ip6_tnl_4rd_parm ip4rd; /* 4rd parameters for the tunnel */
+#endif
 	int err_count;
 	unsigned long err_time;
 
diff --git a/include/uapi/linux/if.h b/include/uapi/linux/if.h
--- a/include/uapi/linux/if.h
+++ b/include/uapi/linux/if.h
@@ -104,6 +104,7 @@ enum net_device_flags {
 	IFF_DORMANT			= 1<<17, /* volatile */
 	IFF_ECHO			= 1<<18, /* volatile */
 #endif /* __UAPI_DEF_IF_NET_DEVICE_FLAGS_LOWER_UP_DORMANT_ECHO */
+	IFF_WIFI_OFLD		= 1<<19, /*Offload interface */
 };
 #endif /* __UAPI_DEF_IF_NET_DEVICE_FLAGS_LOWER_UP_DORMANT_ECHO != 0 || __UAPI_DEF_IF_NET_DEVICE_FLAGS != 0 */
 
@@ -133,6 +134,8 @@ enum net_device_flags {
 #define IFF_ECHO			IFF_ECHO
 #endif /* __UAPI_DEF_IF_NET_DEVICE_FLAGS_LOWER_UP_DORMANT_ECHO */
 
+#define IFF_WIFI_OFLD		IFF_WIFI_OFLD
+
 #define IFF_VOLATILE	(IFF_LOOPBACK|IFF_POINTOPOINT|IFF_BROADCAST|IFF_ECHO|\
 		IFF_MASTER|IFF_SLAVE|IFF_RUNNING|IFF_LOWER_UP|IFF_DORMANT)
 
diff --git a/include/uapi/linux/if_arp.h b/include/uapi/linux/if_arp.h
--- a/include/uapi/linux/if_arp.h
+++ b/include/uapi/linux/if_arp.h
@@ -29,6 +29,7 @@
 /* ARP protocol HARDWARE identifiers. */
 #define ARPHRD_NETROM	0		/* from KA9Q: NET/ROM pseudo	*/
 #define ARPHRD_ETHER 	1		/* Ethernet 10Mbps		*/
+#define ARPHRD_IPV6_IPV6_TUNNEL		ARPHRD_ETHER
 #define	ARPHRD_EETHER	2		/* Experimental Ethernet	*/
 #define	ARPHRD_AX25	3		/* AX.25 Level 2		*/
 #define	ARPHRD_PRONET	4		/* PROnet token ring		*/
diff --git a/include/uapi/linux/if_tunnel.h b/include/uapi/linux/if_tunnel.h
--- a/include/uapi/linux/if_tunnel.h
+++ b/include/uapi/linux/if_tunnel.h
@@ -22,6 +22,16 @@
 #define SIOCDEL6RD      (SIOCDEVPRIVATE + 10)
 #define SIOCCHG6RD      (SIOCDEVPRIVATE + 11)
 
+#define SIOCGET4RD		(SIOCDEVPRIVATE + 12)
+#define SIOCADD4RD		(SIOCDEVPRIVATE + 13)
+#define SIOCDEL4RD		(SIOCDEVPRIVATE + 14)
+#define SIOCCHG4RD		(SIOCDEVPRIVATE + 15)
+//#define COMCERTO_ETHERIPV4
+
+/* SIOCCHG4RD is currently not used and even if it were, these two tunnels would never co-inside*/
+/* MSPD Added */
+#define SIOCISETHIPV4TUNNEL  (SIOCDEVPRIVATE + 15)
+
 #define GRE_CSUM	__cpu_to_be16(0x8000)
 #define GRE_ROUTING	__cpu_to_be16(0x4000)
 #define GRE_KEY		__cpu_to_be16(0x2000)
@@ -114,6 +124,21 @@ struct ip_tunnel_6rd {
 	__u16			relay_prefixlen;
 };
 
+
+/* ip6 tnl 4rd parm -start  */
+struct ip6_tnl_4rd {
+	__be32                  prefix;
+	struct in6_addr         relay_prefix;
+	struct in6_addr         relay_suffix;
+	__u16                   prefixlen;
+	__u16                   relay_prefixlen;
+	__u16                   relay_suffixlen;
+	__u16                   psid_offsetlen;
+	__u16                   eabit_len;
+	__u16                   entry_num;
+};
+/* ip6 tnl 4rd parm -end  */
+
 enum {
 	IFLA_GRE_UNSPEC,
 	IFLA_GRE_LINK,
diff --git a/include/uapi/linux/ip6_tunnel.h b/include/uapi/linux/ip6_tunnel.h
--- a/include/uapi/linux/ip6_tunnel.h
+++ b/include/uapi/linux/ip6_tunnel.h
@@ -53,4 +53,50 @@ struct ip6_tnl_parm2 {
 	__be32			o_key;
 };
 
+struct ip6_4rd_map_msg {
+	__u32 reset;
+	__u32 ifindex;
+	__be32 prefix;
+	__u16 prefixlen;
+	struct in6_addr relay_prefix;
+	struct in6_addr relay_suffix;
+	__u16 relay_prefixlen;
+	__u16 relay_suffixlen;
+	__u16 psid_offsetlen;
+	__u16 eabit_len;
+	__u16 entry_num;
+};
+
+struct ip6_tnl_4rd_map_rule {
+	__be32 prefix;
+	__u16 prefixlen;
+	struct in6_addr relay_prefix;
+	struct in6_addr relay_suffix;
+	__u16 relay_prefixlen;
+	__u16 relay_suffixlen;
+	__u16 psid_offsetlen;
+	__u16 eabit_len;
+	__u16 entry_num;
+	struct list_head mr_list;
+};
+
+#ifdef __KERNEL__
+struct ip6_tnl_4rd_parm {
+	__be32 prefix;
+	struct in6_addr relay_prefix;
+	struct in6_addr relay_suffix;
+	__u16 prefixlen;
+	__u16 relay_prefixlen;
+	__u16 relay_suffixlen;
+	__be32 laddr4;
+	__u16 port_set_id;
+	__u16 port_set_id_len;
+	__u16 psid_offsetlen;
+	__u16 eabit_len;
+
+	struct list_head map_list;
+	rwlock_t map_lock;
+};
 #endif
+#endif
+
diff --git a/include/uapi/linux/rtnetlink.h b/include/uapi/linux/rtnetlink.h
--- a/include/uapi/linux/rtnetlink.h
+++ b/include/uapi/linux/rtnetlink.h
@@ -159,6 +159,14 @@ enum {
 
 	RTM_NEWCACHEREPORT = 96,
 #define RTM_NEWCACHEREPORT RTM_NEWCACHEREPORT
+/* 4RD tunnel support */
+	RTM_NEW4RD = 97,
+#define RTM_NEW4RD		RTM_NEW4RD
+	RTM_DEL4RD,
+#define RTM_DEL4RD		RTM_DEL4RD
+	RTM_GET4RD,
+#define RTM_GET4RD		RTM_GET4RD
+
 
 	RTM_NEWCHAIN = 100,
 #define RTM_NEWCHAIN RTM_NEWCHAIN
diff --git a/net/core/dev.c b/net/core/dev.c
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -176,6 +176,10 @@ static int call_netdevice_notifiers_exta
 					   struct net_device *dev,
 					   struct netlink_ext_ack *extack);
 
+#if defined(CONFIG_CPE_FAST_PATH)
+static fp_iface_stats_get fast_path_stats_get;
+#endif
+
 static DEFINE_MUTEX(ifalias_mutex);
 
 /* protects napi_hash addition/deletion and napi_gen_id */
@@ -4032,10 +4036,17 @@ static struct sk_buff *validate_xmit_skb
 			skb = segs;
 		}
 	} else {
-		if (skb_needs_linearize(skb, features) &&
-		    __skb_linearize(skb))
-			goto out_kfree_skb;
-
+		// linearise only if ipsec policy is not selected
+#if defined(CONFIG_INET_IPSEC_OFFLOAD) || defined(CONFIG_INET6_IPSEC_OFFLOAD)
+		if(!(skb->ipsec_offload))
+#endif // defined(CONFIG_INET_IPSEC_OFFLOAD) || defined(CONFIG_INET6_IPSEC_OFFLOAD)
+		{
+			if (skb_needs_linearize(skb, features))
+			{
+				if (__skb_linearize(skb))
+					goto out_kfree_skb;
+			}
+		}
 		/* If packet is not checksummed and device does not
 		 * support checksumming for this protocol, complete
 		 * checksumming here.
@@ -4832,6 +4843,40 @@ out:
 }
 EXPORT_SYMBOL(__dev_queue_xmit);
 
+#if defined(CONFIG_CPE_FAST_PATH)
+/* WiFi IPSEC offload hook - allows cdx to intercept packets for IPSEC
+ * processing when the packet is transmitted on a wifi interface.
+ */
+dpaa_wifi_xmit_local_hook_t dpaa_wifi_xmit_local_ipsec_handler;
+EXPORT_SYMBOL(dpaa_wifi_xmit_local_ipsec_handler);
+
+/* Register a hook function for IPSEC offload on wifi interfaces */
+int dpa_register_wifi_xmit_local_hook(dpaa_wifi_xmit_local_hook_t hookfn)
+{
+	if (dpaa_wifi_xmit_local_ipsec_handler) {
+		pr_warn("%s: hook already registered\n", __func__);
+		return -1;
+	}
+	dpaa_wifi_xmit_local_ipsec_handler = hookfn;
+	return 0;
+}
+EXPORT_SYMBOL(dpa_register_wifi_xmit_local_hook);
+
+/* Unregister the IPSEC offload hook */
+void dpa_unregister_wifi_xmit_local_hook(void)
+{
+	dpaa_wifi_xmit_local_ipsec_handler = NULL;
+}
+EXPORT_SYMBOL(dpa_unregister_wifi_xmit_local_hook);
+
+/* Original dev_queue_xmit - called when wifi hook is not applicable */
+int original_dev_queue_xmit(struct sk_buff *skb)
+{
+	return __dev_queue_xmit(skb, NULL);
+}
+EXPORT_SYMBOL(original_dev_queue_xmit);
+#endif
+
 int __dev_direct_xmit(struct sk_buff *skb, u16 queue_id)
 {
 	struct net_device *dev = skb->dev;
@@ -5900,6 +5945,14 @@ static int __netif_receive_skb_core(stru
 
 	trace_netif_receive_skb(skb);
 
+#ifdef CONFIG_CPE_FAST_PATH
+	/* ifindex of device we arrived on,now skb->skb_iif
+	always tracks skb->dev */
+	if (!skb->iif_index)
+		skb->iif_index = skb->dev->ifindex;
+	if(!skb->underlying_iif)
+		skb->underlying_iif = skb->dev->ifindex;
+#endif
 	orig_dev = skb->dev;
 
 	skb_reset_network_header(skb);
@@ -11803,10 +11856,28 @@ struct rtnl_link_stats64 *dev_get_stats(
 			storage->rx_otherhost_dropped += READ_ONCE(core_stats->rx_otherhost_dropped);
 		}
 	}
+#if defined(CONFIG_CPE_FAST_PATH)
+	if (fast_path_stats_get)
+		fast_path_stats_get(dev, storage);
+#endif
 	return storage;
 }
 EXPORT_SYMBOL(dev_get_stats);
 
+#if defined(CONFIG_CPE_FAST_PATH)
+void dev_fp_stats_get_register(fp_iface_stats_get func)
+{
+	fast_path_stats_get = func;
+}
+EXPORT_SYMBOL(dev_fp_stats_get_register);
+
+void dev_fp_stats_get_deregister(void)
+{
+	fast_path_stats_get = NULL;
+}
+EXPORT_SYMBOL(dev_fp_stats_get_deregister);
+#endif
+
 /**
  *	dev_fetch_sw_netstats - get per-cpu network device statistics
  *	@s: place to store stats
diff --git a/net/core/gro.c b/net/core/gro.c
--- a/net/core/gro.c
+++ b/net/core/gro.c
@@ -653,6 +653,11 @@ static void napi_reuse_skb(struct napi_s
 	__vlan_hwaccel_clear_tag(skb);
 	skb->dev = napi->dev;
 	skb->skb_iif = 0;
+#ifdef CONFIG_CPE_FAST_PATH
+	skb->iif_index = 0;
+	skb->underlying_iif = 0;
+	skb->underlying_vlan_tci = 0;
+#endif
 
 	/* eth_type_trans() assumes pkt_type is PACKET_HOST */
 	skb->pkt_type = PACKET_HOST;
diff --git a/net/core/rtnetlink.c b/net/core/rtnetlink.c
--- a/net/core/rtnetlink.c
+++ b/net/core/rtnetlink.c
@@ -978,9 +978,17 @@ int rtnetlink_put_metrics(struct sk_buff
 	struct nlattr *mx;
 	int i, valid = 0;
 
+#ifndef CONFIG_CPE_FAST_PATH
+	/*By default this metric is zero for all routes unless mtu is 
+		included while adding route.Due to this mtu is coming as zero 
+		when cmm gets the route through netlink messge.So Added condtion 
+		to bypass this code so that cmm gets right mtu value even if route
+		is added without including mtu
+					*/
 	/* nothing is dumped for dst_default_metrics, so just skip the loop */
 	if (metrics == dst_default_metrics.metrics)
 		return 0;
+#endif
 
 	mx = nla_nest_start_noflag(skb, RTA_METRICS);
 	if (mx == NULL)
@@ -1010,6 +1018,17 @@ int rtnetlink_put_metrics(struct sk_buff
 			}
 			valid++;
 		}
+#ifdef CONFIG_CPE_FAST_PATH
+		else if (i  == RTAX_MTU - 1){
+			struct dst_entry *dst = skb_dst(skb);
+			if(dst)
+			{
+				if (nla_put_u32(skb, i + 1, dst_mtu(dst)))
+					goto nla_put_failure;
+					valid++;
+			}
+		}
+#endif	
 	}
 
 	if (!valid) {
@@ -4487,6 +4506,9 @@ void rtmsg_ifinfo(int type, struct net_d
 	rtmsg_ifinfo_event(type, dev, change, rtnl_get_event(0), flags,
 			   NULL, 0, portid, nlh);
 }
+#if defined(CONFIG_CPE_FAST_PATH)
+EXPORT_SYMBOL_GPL(rtmsg_ifinfo);
+#endif
 
 void rtmsg_ifinfo_newnet(int type, struct net_device *dev, unsigned int change,
 			 gfp_t flags, int *new_nsid, int new_ifindex)
diff --git a/net/ipv4/Kconfig b/net/ipv4/Kconfig
--- a/net/ipv4/Kconfig
+++ b/net/ipv4/Kconfig
@@ -392,6 +392,12 @@ config INET_ESPINTCP
 
 	  If unsure, say N.
 
+config INET_IPSEC_OFFLOAD
+	bool "IPsec Fast Path Processing offload"
+	depends on (INET_ESP || INET_AH) && CPE_FAST_PATH
+	help
+	  Support for IPsec Fast Path offload.
+
 config INET_IPCOMP
 	tristate "IP: IPComp transformation"
 	select INET_XFRM_TUNNEL
diff --git a/net/ipv4/ip_output.c b/net/ipv4/ip_output.c
--- a/net/ipv4/ip_output.c
+++ b/net/ipv4/ip_output.c
@@ -105,6 +105,17 @@ int __ip_local_out(struct net *net, stru
 
 	IP_INC_STATS(net, IPSTATS_MIB_OUTREQUESTS);
 
+#if defined(CONFIG_INET_IPSEC_OFFLOAD)
+	/*
+	 * The tunnel header is not added in slow path and packet may be ipv6
+	 * (IPv6 traffic and IPv4 IPSec tunnel) in case IPv6 over IPv4 IPsec
+	 * tunnel. When it assumes as IPv4 and accessing IP header from SKB
+	 * causing invalid accesses it leading to kernel panic.
+	 * Avoiding ip header checks.
+	 */
+	if((skb->ipsec_offload) && (iph->version == 6))
+		goto sendout;
+#endif /* endif for CONFIG_INET_IPSEC_OFFLOAD */
 	iph_set_totlen(iph, skb->len);
 	ip_send_check(iph);
 
@@ -115,8 +126,19 @@ int __ip_local_out(struct net *net, stru
 	if (unlikely(!skb))
 		return 0;
 
+#if defined(CONFIG_INET_IPSEC_OFFLOAD)
+sendout:
+#endif
 	skb->protocol = htons(ETH_P_IP);
 
+#if defined(CONFIG_INET_IPSEC_OFFLOAD)
+	if(skb->ipsec_offload)
+	{
+		dst_output(net, sk, skb);
+		return 0;
+	}
+	else
+#endif /* endif for CONFIG_INET_IPSEC_OFFLOAD */
 	return nf_hook(NFPROTO_IPV4, NF_INET_LOCAL_OUT,
 		       net, sk, skb, NULL, skb_dst_dev(skb),
 		       dst_output);
@@ -309,7 +331,11 @@ static int __ip_finish_output(struct net
 	if (skb_is_gso(skb))
 		return ip_finish_output_gso(net, sk, skb, mtu);
 
-	if (skb->len > mtu || IPCB(skb)->frag_max_size)
+	if (
+#if defined(CONFIG_INET_IPSEC_OFFLOAD)
+		(skb->ipsec_offload == 0) &&
+#endif
+		(skb->len > mtu || IPCB(skb)->frag_max_size))
 		return ip_fragment(net, sk, skb, mtu, ip_finish_output2);
 
 	return ip_finish_output2(net, sk, skb);
@@ -419,6 +445,13 @@ int ip_mc_output(struct net *net, struct
 				ip_mc_finish_output);
 	}
 
+#if defined(CONFIG_INET_IPSEC_OFFLOAD)
+	/* Bypass invoking post routing hooks since the tunnel header and ESP
+	 * processing is not done in slow path for IPSec offloaded cases
+	 */
+	if (skb->ipsec_offload)
+		return ip_finish_output(net, sk, skb);
+#endif
 	return NF_HOOK_COND(NFPROTO_IPV4, NF_INET_POST_ROUTING,
 			    net, sk, skb, NULL, skb->dev,
 			    ip_finish_output,
@@ -559,6 +592,9 @@ static void ip_copy_metadata(struct sk_b
 	skb_dst_copy(to, from);
 	to->dev = from->dev;
 	to->mark = from->mark;
+#if defined(CONFIG_CPE_FAST_PATH)
+	to->qosmark = from->qosmark;
+#endif
 
 	skb_copy_hash(to, from);
 
diff --git a/net/ipv4/ip_tunnel.c b/net/ipv4/ip_tunnel.c
--- a/net/ipv4/ip_tunnel.c
+++ b/net/ipv4/ip_tunnel.c
@@ -439,6 +439,9 @@ int ip_tunnel_rcv(struct ip_tunnel *tunn
 		skb->protocol = eth_type_trans(skb, tunnel->dev);
 		skb_postpull_rcsum(skb, eth_hdr(skb), ETH_HLEN);
 	} else {
+#ifdef CONFIG_CPE_FAST_PATH
+		skb->underlying_iif = skb->dev->ifindex;
+#endif
 		skb->dev = tunnel->dev;
 	}
 
diff --git a/net/ipv6/Kconfig b/net/ipv6/Kconfig
--- a/net/ipv6/Kconfig
+++ b/net/ipv6/Kconfig
@@ -103,6 +103,12 @@ config INET6_ESPINTCP
 
 	  If unsure, say N.
 
+config INET6_IPSEC_OFFLOAD
+	bool "IPsec IPv6 Fast Path Processing offload"
+	depends on (INET6_ESP && CPE_FAST_PATH)
+	help
+	  Support for IPsec IPv6 Fast Path offload.
+
 config INET6_IPCOMP
 	tristate "IPv6: IPComp transformation"
 	select INET6_XFRM_TUNNEL
diff --git a/net/ipv6/ip6_output.c b/net/ipv6/ip6_output.c
--- a/net/ipv6/ip6_output.c
+++ b/net/ipv6/ip6_output.c
@@ -80,6 +80,13 @@ static int ip6_finish_output2(struct net
 
 	hdr = ipv6_hdr(skb);
 	daddr = &hdr->daddr;
+#if defined(CONFIG_INET6_IPSEC_OFFLOAD)
+	/* For IPv4 over IPv6 IPsec tunnel cases, just send the packet out
+	 * since the packet is IPv4
+	 */
+	if((skb->ipsec_offload) && (hdr->version == 4))
+		goto sendout;
+#endif /* endif for CONFIG_INET6_IPSEC_OFFLOAD */
 	if (ipv6_addr_is_multicast(daddr)) {
 		if (!(dev->flags & IFF_LOOPBACK) && sk_mc_loop(sk) &&
 		    ((mroute6_is_socket(net, skb) &&
@@ -111,6 +118,9 @@ static int ip6_finish_output2(struct net
 		}
 	}
 
+#if defined(CONFIG_INET6_IPSEC_OFFLOAD)
+sendout:
+#endif /* endif for CONFIG_INET6_IPSEC_OFFLOAD */
 	if (lwtunnel_xmit_redirect(dst->lwtstate)) {
 		int res = lwtunnel_xmit(skb);
 
@@ -234,6 +244,13 @@ int ip6_output(struct net *net, struct s
 	skb->protocol = htons(ETH_P_IPV6);
 	rcu_read_lock();
 	dev = dst_dev_rcu(dst);
+#if defined(CONFIG_INET6_IPSEC_OFFLOAD)
+	/* Bypass invoking post routing hooks since the tunnel header and ESP
+	 * processing is not done in slow path for IPSec offloaded cases
+	 */
+	if (skb->ipsec_offload)
+		return ip6_finish_output(net, sk, skb);
+#endif
 	idev = ip6_dst_idev(dst);
 	skb->dev = dev;
 
@@ -697,6 +714,9 @@ static void ip6_copy_metadata(struct sk_
 	skb_dst_set(to, dst_clone(skb_dst(from)));
 	to->dev = from->dev;
 	to->mark = from->mark;
+#if defined(CONFIG_CPE_FAST_PATH)
+	to->qosmark = from->qosmark;
+#endif
 
 	skb_copy_hash(to, from);
 
diff --git a/net/ipv6/ip6_tunnel.c b/net/ipv6/ip6_tunnel.c
--- a/net/ipv6/ip6_tunnel.c
+++ b/net/ipv6/ip6_tunnel.c
@@ -19,6 +19,7 @@
 #include <linux/capability.h>
 #include <linux/errno.h>
 #include <linux/types.h>
+#include <linux/version.h>
 #include <linux/sockios.h>
 #include <linux/icmp.h>
 #include <linux/if.h>
@@ -76,6 +77,505 @@ static u32 HASH(const struct in6_addr *a
 	return hash_32(hash, IP6_TUNNEL_HASH_SIZE_SHIFT);
 }
 
+#ifdef CONFIG_CPE_4RD_TUNNEL
+#define for_each_ip6_tunnel_rcu(start) \
+	for (t = rcu_dereference(start); t; t = rcu_dereference(t->next))
+
+/*
+ * 4RD Tunnel Support Functions
+ * Locking: hash tables are protected by RCU and RTNL
+ */
+static struct kmem_cache *mr_kmem __read_mostly;
+int mr_kmem_alloced = 0;
+
+static inline size_t ip6_4rd_nlmsg_size(void)
+{
+	return NLMSG_ALIGN(sizeof(struct ip6_4rd_map_msg));
+}
+
+static int ip6_4rd_fill_node(struct sk_buff *skb, struct ip6_tnl_4rd_map_rule *mr,
+		u32 pid, u32 seq, int type, unsigned int flags, int reset,
+		unsigned int ifindex)
+{
+	struct ip6_4rd_map_msg *mr_msg;
+	struct nlmsghdr *nlh;
+
+	nlh = nlmsg_put(skb, pid, seq, type, sizeof(*mr_msg), flags);
+	if (nlh == NULL)
+		return -EMSGSIZE;
+
+	mr_msg = nlmsg_data(nlh);
+	if (reset) {
+		memset(mr_msg, 0, sizeof(*mr_msg));
+		mr_msg->reset = 1;
+		mr_msg->ifindex = ifindex;
+	} else {
+		memset(mr_msg, 0, sizeof(*mr_msg));
+		mr_msg->prefix = mr->prefix;
+		mr_msg->prefixlen = mr->prefixlen;
+		mr_msg->relay_prefix = mr->relay_prefix;
+		mr_msg->relay_suffix = mr->relay_suffix;
+		mr_msg->relay_prefixlen = mr->relay_prefixlen;
+		mr_msg->relay_suffixlen = mr->relay_suffixlen;
+		mr_msg->psid_offsetlen = mr->psid_offsetlen;
+		mr_msg->eabit_len = mr->eabit_len;
+		mr_msg->entry_num = mr->entry_num;
+		mr_msg->ifindex = ifindex;
+	}
+	nlmsg_end(skb, nlh);
+	return nlh->nlmsg_len;
+}
+
+static void ip6_4rd_notify(int event, struct ip6_tnl_4rd_map_rule *mr,
+		struct net_device *dev, int reset)
+{
+	struct sk_buff *skb;
+	struct net *net = dev_net(dev);
+	int err;
+
+	err = -ENOBUFS;
+	skb = nlmsg_new(ip6_4rd_nlmsg_size(), gfp_any());
+	if (skb == NULL)
+		goto errout;
+
+	err = ip6_4rd_fill_node(skb, mr, 0, 0, event, 0, reset, dev->ifindex);
+	if (err < 0) {
+		WARN_ON(err == -EMSGSIZE);
+		kfree_skb(skb);
+		goto errout;
+	}
+	rtnl_notify(skb, net, 0, RTNLGRP_IPV6_IFADDR, NULL, gfp_any());
+	return;
+errout:
+	if (err < 0)
+		rtnl_set_sk_err(net, RTNLGRP_IPV6_IFADDR, err);
+}
+
+static inline void
+ip6_tnl_4rd_mr_destroy(char *f, struct ip6_tnl_4rd_map_rule *mr)
+{
+	list_del(&mr->mr_list);
+	kmem_cache_free(mr_kmem, mr);
+	--mr_kmem_alloced;
+}
+
+static int
+ip6_tnl_4rd_mr_create(struct ip6_tnl_4rd *ip4rd, struct ip6_tnl_4rd_parm *parm,
+		struct net_device *dev)
+{
+	struct ip6_tnl_4rd_map_rule *mr;
+	int err = 0;
+
+	write_lock_bh(&parm->map_lock);
+	list_for_each_entry(mr, &parm->map_list, mr_list) {
+		if (mr->entry_num == ip4rd->entry_num) {
+			pr_debug("ip6_tnl_4rd_mr_create: map rule found update\n");
+			mr->prefix = ip4rd->prefix;
+			mr->relay_prefix = ip4rd->relay_prefix;
+			mr->relay_suffix = ip4rd->relay_suffix;
+			mr->prefixlen = ip4rd->prefixlen;
+			mr->relay_prefixlen = ip4rd->relay_prefixlen;
+			mr->relay_suffixlen = ip4rd->relay_suffixlen;
+			mr->psid_offsetlen = ip4rd->psid_offsetlen;
+			mr->eabit_len = ip4rd->eabit_len;
+			mr->entry_num = ip4rd->entry_num;
+			goto out;
+		}
+	}
+
+	mr = kmem_cache_alloc(mr_kmem, GFP_KERNEL);
+	if (!mr) {
+		pr_info("ip6_tnl_4rd_mr_create: kmem_cache_alloc fail\n");
+		err = -1;
+		goto out;
+	}
+
+	mr->prefix = ip4rd->prefix;
+	mr->relay_prefix = ip4rd->relay_prefix;
+	mr->relay_suffix = ip4rd->relay_suffix;
+	mr->prefixlen = ip4rd->prefixlen;
+	mr->relay_prefixlen = ip4rd->relay_prefixlen;
+	mr->relay_suffixlen = ip4rd->relay_suffixlen;
+	mr->psid_offsetlen = ip4rd->psid_offsetlen;
+	mr->eabit_len = ip4rd->eabit_len;
+	mr->entry_num = ip4rd->entry_num;
+
+	++mr_kmem_alloced;
+	list_add_tail(&mr->mr_list, &parm->map_list);
+
+out:
+	ip6_4rd_notify(RTM_NEW4RD, mr, dev, 0);
+	write_unlock_bh(&parm->map_lock);
+	return err;
+}
+
+static void
+ip6_tnl_4rd_mr_delete_all(struct ip6_tnl_4rd_parm *parm, struct net_device *dev)
+{
+	struct ip6_tnl_4rd_map_rule *mr, *mr_rule;
+
+	write_lock_bh(&parm->map_lock);
+	list_for_each_entry_safe(mr, mr_rule, &parm->map_list, mr_list) {
+		ip6_tnl_4rd_mr_destroy("all", mr);
+	}
+	ip6_4rd_notify(RTM_DEL4RD, mr, dev, 1);
+	write_unlock_bh(&parm->map_lock);
+}
+
+static int
+ip6_tnl_4rd_mr_delete(__u16 entry_num, struct ip6_tnl_4rd_parm *parm,
+		struct net_device *dev)
+{
+	struct ip6_tnl_4rd_map_rule *mr, *mr_rule;
+	int err = -1;
+
+	write_lock_bh(&parm->map_lock);
+	list_for_each_entry_safe(mr, mr_rule, &parm->map_list, mr_list) {
+		if (mr->entry_num == entry_num) {
+			pr_debug("ip6_tnl_4rd_mr_delete: map rule found delete\n");
+			ip6_tnl_4rd_mr_destroy("one", mr);
+			err = 0;
+			break;
+		}
+	}
+	ip6_4rd_notify(RTM_DEL4RD, mr, dev, 0);
+	write_unlock_bh(&parm->map_lock);
+	return err;
+}
+
+static void
+ip6_tnl_4rd_mr_show(struct ip6_tnl_4rd_parm *parm)
+{
+	struct ip6_tnl_4rd_map_rule *mr;
+
+	pr_debug("-- 4rd mapping rule list\n");
+	pr_debug("-- entry num = %d\n", mr_kmem_alloced);
+
+	read_lock(&parm->map_lock);
+	list_for_each_entry(mr, &parm->map_list, mr_list) {
+		pr_debug("%03d : %pI4/%02d %pI6c/%03d %pI6c/%03d eabit:%03d offset:%03d\n",
+			mr->entry_num,
+			&mr->prefix,
+			mr->prefixlen,
+			&mr->relay_prefix,
+			mr->relay_prefixlen,
+			&mr->relay_suffix,
+			mr->relay_suffixlen,
+			mr->eabit_len,
+			mr->psid_offsetlen);
+	}
+	read_unlock(&parm->map_lock);
+}
+
+static int
+ip6_tnl_4rd_modify_daddr(struct in6_addr *daddr6, __be32 daddr4, __be16 dport4,
+		struct ip6_tnl_4rd_map_rule *mr)
+{
+	int i, pbw0, pbi0, pbi1;
+	__u32 daddr[4];
+	__u32 port_set_id = 0;
+	__u32 mask;
+	__u32 da = ntohl(daddr4);
+	__u16 dp = ntohs(dport4);
+	__u32 diaddr[4];
+	int port_set_id_len = (mr->eabit_len) - (32 - mr->prefixlen);
+
+	if (port_set_id_len < 0) {
+		pr_debug("ip6_tnl_4rd_modify_daddr: PSID length ERROR %d\n",
+			port_set_id_len);
+		return -1;
+	}
+
+	if (port_set_id_len > 0) {
+		mask = 0xffffffff >> (32 - port_set_id_len);
+		port_set_id = (dp >> (16 - mr->psid_offsetlen - port_set_id_len) & mask);
+	}
+
+	for (i = 0; i < 4; ++i)
+		daddr[i] = ntohl(mr->relay_prefix.s6_addr32[i])
+			| ntohl(mr->relay_suffix.s6_addr32[i]);
+
+	if (mr->prefixlen < 32) {
+		pbw0 = mr->relay_prefixlen >> 5;
+		pbi0 = mr->relay_prefixlen & 0x1f;
+		daddr[pbw0] |= (da << mr->prefixlen) >> pbi0;
+		pbi1 = pbi0 - mr->prefixlen;
+		if (pbi1 > 0)
+			daddr[pbw0+1] |= da << (32 - pbi1);
+	}
+	if (port_set_id_len > 0) {
+		pbw0 = (mr->relay_prefixlen + 32 - mr->prefixlen) >> 5;
+		pbi0 = (mr->relay_prefixlen + 32 - mr->prefixlen) & 0x1f;
+		daddr[pbw0] |= (port_set_id << (32 - port_set_id_len)) >> pbi0;
+		pbi1 = pbi0 - (32 - port_set_id_len);
+		if (pbi1 > 0)
+			daddr[pbw0+1] |= port_set_id << (32 - pbi1);
+	}
+
+	memset(diaddr, 0, sizeof(diaddr));
+	diaddr[2] = (da >> 8);
+	diaddr[3] = (da << 24);
+	diaddr[3] |= (port_set_id << 8);
+
+	for (i = 0; i < 4; ++i)
+		daddr[i] = daddr[i] | diaddr[i];
+
+	for (i = 0; i < 4; ++i)
+		daddr6->s6_addr32[i] = htonl(daddr[i]);
+
+	pr_debug("ip6_tnl_4rd_modify_daddr: %08x %08x %08x %08x  PSID:%04x\n",
+		daddr[0], daddr[1], daddr[2], daddr[3], port_set_id);
+
+	return 0;
+}
+
+/**
+ * ip6_tnl_4rd_rcv_helper - validate received 4RD packet
+ *   @skb: received socket buffer
+ *   @t: tunnel device
+ **/
+static int
+ip6_tnl_4rd_rcv_helper(struct sk_buff *skb, struct ip6_tnl *t)
+{
+	int err = 0;
+	struct iphdr *iph;
+
+	iph = ip_hdr(skb);
+
+	switch (iph->protocol) {
+	case IPPROTO_TCP:
+	case IPPROTO_UDP:
+	case IPPROTO_ICMP:
+	case IPPROTO_GRE:
+		break;
+	default:
+		err = -1;
+		break;
+	}
+
+	return err;
+}
+
+static int
+ip6_tnl_4rd_xmit_helper(struct sk_buff *skb, struct flowi6 *fl6,
+		struct ip6_tnl *t)
+{
+	int err = 0;
+	struct iphdr *iph, *icmpiph;
+	__be16 *idp;
+	struct tcphdr *tcph, *icmptcph;
+	struct udphdr *udph, *icmpudph;
+	struct icmphdr *icmph;
+	__u32 mask;
+	__be16 *sportp = NULL;
+	__be32 daddr;
+	__be16 dport;
+	u8 *ptr;
+	int no_dst_chg = 0;
+	struct ip6_tnl_4rd_map_rule *mr, *mr_tmp;
+	int mr_prefixlen;
+	int count;
+
+	iph = ip_hdr(skb);
+
+	daddr = iph->daddr;
+	idp = &iph->id;
+
+	ptr = (u8 *)iph;
+	ptr += iph->ihl * 4;
+	switch (iph->protocol) {
+	case IPPROTO_TCP:
+		tcph = (struct tcphdr *)ptr;
+		sportp = &tcph->source;
+		dport = tcph->dest;
+		break;
+	case IPPROTO_UDP:
+		udph = (struct udphdr *)ptr;
+		sportp = &udph->source;
+		dport = udph->dest;
+		break;
+	case IPPROTO_ICMP:
+		icmph = (struct icmphdr *)ptr;
+		switch (icmph->type) {
+		case ICMP_DEST_UNREACH:
+		case ICMP_SOURCE_QUENCH:
+		case ICMP_REDIRECT:
+		case ICMP_TIME_EXCEEDED:
+		case ICMP_PARAMETERPROB:
+			ptr = (u8 *)icmph;
+			ptr += sizeof(struct icmphdr);
+			icmpiph = (struct iphdr *)ptr;
+			if (ntohs(iph->tot_len) < icmpiph->ihl * 4 + 12) {
+				err = -1;
+				goto out;
+			}
+			daddr = icmpiph->saddr;
+			ptr += icmpiph->ihl * 4;
+			switch (icmpiph->protocol) {
+			case IPPROTO_TCP:
+				icmptcph = (struct tcphdr *)ptr;
+				sportp = &icmptcph->dest;
+				dport = icmptcph->source;
+				break;
+			case IPPROTO_UDP:
+				icmpudph = (struct udphdr *)ptr;
+				sportp = &icmpudph->dest;
+				dport = icmpudph->source;
+				break;
+			default:
+				err = -1;
+				goto out;
+			}
+			break;
+		default:
+			no_dst_chg = 1;
+			break;
+		}
+		break;
+	default:
+		err = -1;
+		goto out;
+	}
+
+	if (no_dst_chg == 0) {
+		count = 0;
+		mr_prefixlen = 0;
+
+		read_lock(&t->ip4rd.map_lock);
+		list_for_each_entry(mr, &t->ip4rd.map_list, mr_list) {
+			mask = 0xffffffff << (32 - mr->prefixlen);
+			if ((htonl(daddr) & mask) == htonl(mr->prefix)) {
+				if (mr->prefixlen >= mr_prefixlen) {
+					mr_prefixlen = mr->prefixlen;
+					mr_tmp = mr;
+					count++;
+				}
+			}
+		}
+
+		if (count) {
+			err = ip6_tnl_4rd_modify_daddr(&fl6->daddr, daddr, dport, mr_tmp);
+			if (err) {
+				read_unlock(&t->ip4rd.map_lock);
+				goto out;
+			}
+		}
+		read_unlock(&t->ip4rd.map_lock);
+
+		if (sportp && idp)
+			*idp = *sportp;
+	}
+
+	iph->check = 0;
+	iph->check = ip_fast_csum((unsigned char *)iph, iph->ihl);
+
+out:
+	return err;
+}
+
+/**
+ * ip6_tnl_4rd_update_parms - update 4rd parameters
+ *   @t: tunnel to be updated
+ *
+ * Description:
+ *   ip6_tnl_4rd_update_parms() updates 4rd parameters
+ **/
+static void
+ip6_tnl_4rd_update_parms(struct ip6_tnl *t)
+{
+	int pbw0, pbi0, pbi1;
+	__u32 d;
+
+	t->ip4rd.port_set_id_len = t->ip4rd.relay_suffixlen
+		- t->ip4rd.relay_prefixlen
+		- (32 - t->ip4rd.prefixlen);
+	pbw0 = (t->ip4rd.relay_suffixlen - t->ip4rd.port_set_id_len) >> 5;
+	pbi0 = (t->ip4rd.relay_suffixlen - t->ip4rd.port_set_id_len) & 0x1f;
+	d = (ntohl(t->parms.laddr.s6_addr32[pbw0]) << pbi0)
+		>> (32 - t->ip4rd.port_set_id_len);
+	pbi1 = pbi0 - (32 - t->ip4rd.port_set_id_len);
+
+	if (pbi1 > 0)
+		d |= ntohl(t->parms.laddr.s6_addr32[pbw0+1]) >> (32 - pbi1);
+	t->ip4rd.port_set_id = d;
+
+	/* local v4 address */
+	t->ip4rd.laddr4 = t->ip4rd.prefix;
+	pbw0 = t->ip4rd.relay_prefixlen >> 5;
+	pbi0 = t->ip4rd.relay_prefixlen & 0x1f;
+	d = (ntohl(t->parms.laddr.s6_addr32[pbw0]) << pbi0)
+		>> t->ip4rd.prefixlen;
+	pbi1 = pbi0 - t->ip4rd.prefixlen;
+	if (pbi1 > 0)
+		d |= ntohl(t->parms.laddr.s6_addr32[pbw0+1]) >> (32 - pbi1);
+	t->ip4rd.laddr4 |= htonl(d);
+	if (t->ip4rd.port_set_id_len < 0) {
+		d = ntohl(t->ip4rd.laddr4);
+		d &= 0xffffffff << -t->ip4rd.port_set_id_len;
+		t->ip4rd.laddr4 = htonl(d);
+	}
+}
+
+static int inet6_dump4rd_mrule(struct sk_buff *skb, struct netlink_callback *cb)
+{
+	struct net *net = sock_net(skb->sk);
+	unsigned int h, s_h;
+	int s_idx, s_ip_idx;
+	int idx, ip_idx;
+	struct ip6_tnl_4rd_map_rule *mr;
+	int err = 0;
+	struct ip6_tnl *t;
+	struct ip6_tnl_net *ip6n = net_generic(net, ip6_tnl_net_id);
+
+	s_h = cb->args[0];
+	s_idx = idx = cb->args[1];
+	s_ip_idx = ip_idx = cb->args[2];
+
+	rcu_read_lock();
+
+	for (h = s_h; h < IP6_TUNNEL_HASH_SIZE; h++, s_idx = 0) {
+		idx = 0;
+		for_each_ip6_tunnel_rcu(ip6n->tnls_r_l[h]) {
+			if (idx < s_idx)
+				goto cont_tunnel;
+			if (idx > s_idx)
+				s_ip_idx = 0;
+			ip_idx = 0;
+			read_lock(&t->ip4rd.map_lock);
+			list_for_each_entry(mr, &t->ip4rd.map_list, mr_list) {
+				if (ip_idx < s_ip_idx)
+					goto cont_mr;
+
+				err = ip6_4rd_fill_node(skb, mr,
+					NETLINK_CB(cb->skb).portid,
+					cb->nlh->nlmsg_seq, RTM_NEW4RD,
+					NLM_F_MULTI, 0, t->dev->ifindex);
+				if (err < 0) {
+					WARN_ON(err == -EMSGSIZE);
+					kfree_skb(skb);
+					read_unlock(&t->ip4rd.map_lock);
+					goto out;
+				}
+cont_mr:
+				ip_idx++;
+			}
+			read_unlock(&t->ip4rd.map_lock);
+cont_tunnel:
+			idx++;
+		}
+	}
+out:
+	rcu_read_unlock();
+
+	cb->args[0] = h;
+	cb->args[1] = idx;
+	cb->args[2] = ip_idx;
+
+	return skb->len;
+}
+
+#endif /* CONFIG_CPE_4RD_TUNNEL */
+
 static int ip6_tnl_dev_init(struct net_device *dev);
 static void ip6_tnl_dev_setup(struct net_device *dev);
 static struct rtnl_link_ops ip6_link_ops __read_mostly;
@@ -96,8 +596,10 @@ static inline int ip6_tnl_mpls_supported
 	return IS_ENABLED(CONFIG_MPLS);
 }
 
+#ifndef CONFIG_CPE_4RD_TUNNEL
 #define for_each_ip6_tunnel_rcu(start) \
 	for (t = rcu_dereference(start); t; t = rcu_dereference(t->next))
+#endif
 
 /**
  * ip6_tnl_lookup - fetch tunnel matching the end-point addresses
@@ -875,6 +1377,11 @@ static int __ip6_tnl_rcv(struct ip6_tnl
 		}
 	}
 
+#ifdef CONFIG_CPE_4RD_TUNNEL
+	if (ip6_tnl_4rd_rcv_helper(skb, tunnel))
+		goto drop;
+#endif
+
 	dev_sw_netstats_rx_add(tunnel->dev, skb->len);
 
 	skb_scrub_packet(skb, !net_eq(tunnel->net, dev_net(tunnel->dev)));
@@ -1201,10 +1708,27 @@ route_lookup:
 		       IPV6_MIN_MTU : IPV4_MIN_MTU);
 
 	skb_dst_update_pmtu_no_confirm(skb, mtu);
-	if (skb->len - t->tun_hlen - eth_hlen > mtu && !skb_is_gso(skb)) {
-		*pmtu = mtu;
-		err = -EMSGSIZE;
-		goto tx_err_dst_release;
+
+#ifdef CONFIG_CPE_4RD_TUNNEL
+		if (!t->ip4rd.prefix) {   /* 4rd support requires Post Fragmentation, so skb->len could be greater than MTU */ 
+#endif
+			if (skb->len - t->tun_hlen - eth_hlen > mtu && !skb_is_gso(skb)) {
+				*pmtu = mtu;
+				err = -EMSGSIZE;
+				goto tx_err_dst_release;
+			}
+#ifdef CONFIG_CPE_4RD_TUNNEL
+		}                         
+
+	if (t->ip4rd.prefix) {
+		struct iphdr *iph;
+		iph = ip_hdr(skb);
+		hop_limit = iph->ttl;
+	}
+	else
+#endif
+	{
+		hop_limit = t->parms.hop_limit;
 	}
 
 	if (t->err_count > 0) {
@@ -1402,6 +1926,12 @@ ipxip6_tnl_xmit(struct sk_buff *skb, str
 
 	skb_set_inner_ipproto(skb, protocol);
 
+#ifdef CONFIG_CPE_4RD_TUNNEL
+	if (protocol == IPPROTO_IPIP &&
+	    t->ip4rd.prefix && ip6_tnl_4rd_xmit_helper(skb, &fl6, t))
+		return -1;
+#endif
+
 	err = ip6_tnl_xmit(skb, dev, dsfield, &fl6, encap_limit, &mtu,
 			   protocol);
 	if (err != 0) {
@@ -1436,6 +1966,10 @@ ip6_tnl_start_xmit(struct sk_buff *skb,
 
 	switch (skb->protocol) {
 	case htons(ETH_P_IP):
+#ifdef CONFIG_CPE_4RD_TUNNEL
+		if (t->ip4rd.prefix && ip_defrag(skb, IP_DEFRAG_IP6_TNL_4RD))
+			return NETDEV_TX_OK;
+#endif
 		ipproto = IPPROTO_IPIP;
 		break;
 	case htons(ETH_P_IPV6):
@@ -1651,6 +2185,10 @@ ip6_tnl_siocdevprivate(struct net_device
 	struct ip6_tnl *t = netdev_priv(dev);
 	struct net *net = t->net;
 	struct ip6_tnl_net *ip6n = net_generic(net, ip6_tnl_net_id);
+#ifdef CONFIG_CPE_4RD_TUNNEL
+	struct ip6_tnl_4rd ip4rd, *ip4rdp;  
+	struct ip6_tnl_4rd_map_rule *mr;   
+#endif
 
 	memset(&p1, 0, sizeof(p1));
 
@@ -1731,9 +2269,120 @@ ip6_tnl_siocdevprivate(struct net_device
 		err = 0;
 		unregister_netdevice(dev);
 		break;
+#ifdef CONFIG_CPE_4RD_TUNNEL
+	case SIOCADD4RD:
+	case SIOCDEL4RD:
+		err = -EPERM;
+		if (!capable(CAP_NET_ADMIN))
+			goto done;
+
+		err = -EFAULT;
+		if (copy_from_user(&ip4rd, ifr->ifr_ifru.ifru_data, sizeof(ip4rd)))
+			goto done;
+
+		t = netdev_priv(dev);
+
+		if (cmd == SIOCADD4RD) {
+
+			__be32 prefix;
+			struct in6_addr relay_prefix, relay_suffix;
+
+			err = -EINVAL;
+
+			if (ip4rd.relay_suffixlen > 64)
+				goto done;
+
+			if (ip4rd.relay_suffixlen <= ip4rd.relay_prefixlen)
+				goto done;
+
+			prefix = ip4rd.prefix & htonl(0xffffffffUL << (32 - ip4rd.prefixlen));
+			if (prefix != ip4rd.prefix)
+				goto done;
+
+			ipv6_addr_prefix(&relay_prefix, &ip4rd.relay_prefix, ip4rd.relay_prefixlen);
+			if (!ipv6_addr_equal(&relay_prefix, &ip4rd.relay_prefix))
+				goto done;
+
+			ipv6_addr_prefix(&relay_suffix, &ip4rd.relay_suffix, ip4rd.relay_suffixlen);
+			if (!ipv6_addr_equal(&relay_suffix, &ip4rd.relay_suffix))
+				goto done;
+
+
+			err = ip6_tnl_4rd_mr_create(&ip4rd, &t->ip4rd, t->dev); /* modified by MSPD */
+
+			if ( ip4rd.entry_num == 0 ){
+
+				t->ip4rd.prefix = prefix;
+				t->ip4rd.relay_prefix = relay_prefix;
+				t->ip4rd.relay_suffix = relay_suffix;
+				t->ip4rd.prefixlen = ip4rd.prefixlen;
+				t->ip4rd.relay_prefixlen = ip4rd.relay_prefixlen;
+				t->ip4rd.relay_suffixlen = ip4rd.relay_suffixlen;
+				t->ip4rd.psid_offsetlen = ip4rd.psid_offsetlen;
+
+				ip6_tnl_4rd_update_parms(t);
+				ip6_tnl_dst_reset(t);
+				ip6_tnl_link_config(t);
+			}
+			/* DBG */
+			ip6_tnl_4rd_mr_show(&t->ip4rd);
+		}else if(cmd == SIOCDEL4RD){
+			if ( ip4rd.entry_num == 0 ){
+				ip6_tnl_4rd_mr_delete_all(&t->ip4rd, t->dev);
+				t->ip4rd.prefix = 0;
+				memset(&t->ip4rd.relay_prefix, 0, sizeof(t->ip4rd.relay_prefix));
+				memset(&t->ip4rd.relay_suffix, 0, sizeof(t->ip4rd.relay_suffix));
+				t->ip4rd.prefixlen = 0;
+				t->ip4rd.relay_prefixlen = 0;
+				t->ip4rd.relay_suffixlen = 0;
+				t->ip4rd.psid_offsetlen = 0;
+				t->ip4rd.laddr4 = 0;
+				t->ip4rd.port_set_id = t->ip4rd.port_set_id_len = 0;
+
+				ip6_tnl_dst_reset(t);
+				ip6_tnl_link_config(t);
+			}else{
+				err = ip6_tnl_4rd_mr_delete( ip4rd.entry_num , &t->ip4rd, t->dev);  /* modified by MSPD */
+			}
+			/* DBG */
+			ip6_tnl_4rd_mr_show(&t->ip4rd);
+		}else{
+			printk(KERN_ERR "=== ioctl_cmd 0x%x \n",cmd );
+		}
+		err = 0;
+		break;
+	case SIOCGET4RD:
+		t = netdev_priv(dev);
+		ip4rdp = (struct ip6_tnl_4rd *)ifr->ifr_ifru.ifru_data;
+
+		read_lock(&t->ip4rd.map_lock);
+		list_for_each_entry (mr, &t->ip4rd.map_list, mr_list){
+			ip4rd.relay_prefix = mr->relay_prefix;
+			ip4rd.relay_suffix = mr->relay_suffix;
+			ip4rd.prefix = mr->prefix;
+			ip4rd.relay_prefixlen = mr->relay_prefixlen;
+			ip4rd.relay_suffixlen = mr->relay_suffixlen;
+			ip4rd.prefixlen = mr->prefixlen;
+			ip4rd.eabit_len = mr->eabit_len;
+			ip4rd.psid_offsetlen = mr->psid_offsetlen;
+			ip4rd.entry_num = mr->entry_num;
+
+			if (copy_to_user(ip4rdp, &ip4rd, sizeof(ip4rd))) {
+				read_unlock(&t->ip4rd.map_lock);
+				err = -EFAULT;
+			}
+			ip4rdp++;
+		}
+		read_unlock(&t->ip4rd.map_lock);
+		break;
+#endif
+
 	default:
 		err = -EINVAL;
 	}
+#ifdef CONFIG_CPE_4RD_TUNNEL
+done:
+#endif
 	return err;
 }
 
@@ -2344,9 +2993,24 @@ static int __init ip6_tunnel_init(void)
 	if (!ipv6_mod_enabled())
 		return -EOPNOTSUPP;
 
+#ifdef CONFIG_CPE_4RD_TUNNEL
+	mr_kmem = kmem_cache_create("ip6_tnl_4rd_map_rule",
+				sizeof(struct ip6_tnl_4rd_map_rule), 0, SLAB_HWCACHE_ALIGN,
+				NULL);
+	if (!mr_kmem)
+	{
+		err= -ENOMEM; 
+		goto out_pernet;
+	}
+#endif
+
 	err = register_pernet_device(&ip6_tnl_net_ops);
 	if (err < 0)
+#ifdef CONFIG_CPE_4RD_TUNNEL
+		goto out_kmem;
+#else
 		goto out_pernet;
+#endif
 
 	err = xfrm6_tunnel_register(&ip4ip6_handler, AF_INET);
 	if (err < 0) {
@@ -2383,6 +3047,10 @@ out_ip6ip6:
 	xfrm6_tunnel_deregister(&ip4ip6_handler, AF_INET);
 out_ip4ip6:
 	unregister_pernet_device(&ip6_tnl_net_ops);
+#ifdef CONFIG_CPE_4RD_TUNNEL
+out_kmem:
+	kmem_cache_destroy(mr_kmem);
+#endif
 out_pernet:
 	return err;
 }
diff --git a/net/ipv6/netfilter/ip6t_NPT.c b/net/ipv6/netfilter/ip6t_NPT.c
--- a/net/ipv6/netfilter/ip6t_NPT.c
+++ b/net/ipv6/netfilter/ip6t_NPT.c
@@ -11,6 +11,9 @@
 #include <linux/netfilter_ipv6.h>
 #include <linux/netfilter_ipv6/ip6t_NPT.h>
 #include <linux/netfilter/x_tables.h>
+#ifdef CONFIG_CPE_FAST_PATH
+#include <net/netfilter/nf_conntrack.h>
+#endif
 
 static int ip6t_npt_checkentry(const struct xt_tgchk_param *par)
 {
@@ -114,6 +117,19 @@ ip6t_snpt_tg(struct sk_buff *skb, const
 			ip6t_npt_map_pfx(npt, &bounced_hdr->daddr);
 	}
 
+#ifdef CONFIG_CPE_FAST_PATH
+	{
+		struct nf_conn *ct;
+		enum ip_conntrack_info ctinfo;
+		rcu_read_lock();
+		ct = nf_ct_get(skb, &ctinfo);
+		if (ct && ((ctinfo == IP_CT_NEW) || (ctinfo == IP_CT_RELATED))) {
+			memcpy(&ct->tuplehash[IP_CT_DIR_REPLY].tuple.dst.u3.in6,
+			       &ipv6_hdr(skb)->saddr, sizeof(struct in6_addr));
+		}
+		rcu_read_unlock();
+	}
+#endif
 	return XT_CONTINUE;
 }
 
@@ -139,6 +155,19 @@ ip6t_dnpt_tg(struct sk_buff *skb, const
 			ip6t_npt_map_pfx(npt, &bounced_hdr->saddr);
 	}
 
+#ifdef CONFIG_CPE_FAST_PATH
+	{
+		struct nf_conn *ct;
+		enum ip_conntrack_info ctinfo;
+		rcu_read_lock();
+		ct = nf_ct_get(skb, &ctinfo);
+		if (ct && ((ctinfo == IP_CT_NEW) || (ctinfo == IP_CT_RELATED))) {
+			memcpy(&ct->tuplehash[IP_CT_DIR_REPLY].tuple.src.u3.in6,
+			       &ipv6_hdr(skb)->daddr, sizeof(struct in6_addr));
+		}
+		rcu_read_unlock();
+	}
+#endif
 	return XT_CONTINUE;
 }
 
diff --git a/net/ipv6/output_core.c b/net/ipv6/output_core.c
--- a/net/ipv6/output_core.c
+++ b/net/ipv6/output_core.c
@@ -130,7 +130,18 @@ int __ip6_local_out(struct net *net, str
 	len = skb->len - sizeof(struct ipv6hdr);
 	if (len > IPV6_MAXPLEN)
 		len = 0;
+
+#if defined(CONFIG_INET6_IPSEC_OFFLOAD)
+/* Since tunnel header is not added in slow path and route cache entry lookup is changed to perform
+   lookup on tunnel header, for Ipv4 over Ipv6 IPsec tunnel cases,dst_output will be ip6 output 
+   but packet will be IPv4 type. So,this check is introduced to update ipv6 pay_load for the following cases
+   1.IPv6 over Ipv6 IPsec tunnel.
+   2.IPv4 over Ipv6 Ipsec tunnel when software encryption come into play(i.e for packets > 1782 bytes) */
+        if((ipv6_hdr(skb)->version) == 6)
+                ipv6_hdr(skb)->payload_len = htons(len);
+#else
 	ipv6_hdr(skb)->payload_len = htons(len);
+#endif
 	IP6CB(skb)->nhoff = offsetof(struct ipv6hdr, nexthdr);
 
 	/* if egress device is enslaved to an L3 master device pass the
@@ -142,6 +153,14 @@ int __ip6_local_out(struct net *net, str
 
 	skb->protocol = htons(ETH_P_IPV6);
 
+#if defined(CONFIG_INET6_IPSEC_OFFLOAD)
+	if(skb->ipsec_offload)
+	{
+		dst_output(net, sk, skb);
+		return 0;
+	}
+	else
+#endif
 	return nf_hook(NFPROTO_IPV6, NF_INET_LOCAL_OUT,
 		       net, sk, skb, NULL, skb_dst_dev(skb),
 		       dst_output);
diff --git a/net/ipv6/sit.c b/net/ipv6/sit.c
--- a/net/ipv6/sit.c
+++ b/net/ipv6/sit.c
@@ -694,6 +694,9 @@ static int ipip6_rcv(struct sk_buff *skb
 		skb->mac_header = skb->network_header;
 		skb_reset_network_header(skb);
 		IPCB(skb)->flags = 0;
+#ifdef CONFIG_CPE_FAST_PATH
+		skb->underlying_iif = skb->dev->ifindex;
+#endif
 		skb->dev = tunnel->dev;
 
 		if (packet_is_spoofed(skb, iph, tunnel)) {
diff --git a/net/ipv6/udp.c b/net/ipv6/udp.c
--- a/net/ipv6/udp.c
+++ b/net/ipv6/udp.c
@@ -871,7 +871,6 @@ static int udpv6_queue_rcv_one_skb(struc
 
 		/* FALLTHROUGH -- it's a UDP Packet */
 	}
-
 	/*
 	 * UDP-Lite specific tests, ignored on UDP sockets (see net/ipv4/udp.c).
 	 */
