From ecf727ecf72ceb9e3c5ac7d44ac340ee8bd287f5 Mon Sep 17 00:00:00 2001
From: "miha.kralj" <miha.kralj@outlook.com>
Date: Mon, 20 Jul 2026 03:56:42 +0000
Subject: [PATCH] temp: vyos-1x-005-fix_podman_service.patch

---
 src/conf_mode/container.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/conf_mode/container.py b/src/conf_mode/container.py
index 764c26d..b2464f7 100755
--- a/src/conf_mode/container.py
+++ b/src/conf_mode/container.py
@@ -489,7 +489,7 @@ def generate_run_arguments(name, container_config, host_ident):
         name_server = ''
 
     container_base_cmd = f'--detach --interactive --tty --replace {capabilities} {privileged} --cpus {cpu_quota} {sysctl_opt} ' \
-                         f'--memory {memory}m --shm-size {shared_memory}m --memory-swap 0 --restart {restart} --log-driver={log_driver} ' \
+                         f'--memory {memory}m --shm-size {shared_memory}m --restart {restart} --log-driver={log_driver} ' \
                          f'--name {name} {hostname} {device} {port} {name_server} {volume} {tmpfs} {env_opt} {label} {uid} {host_pid} {cgroupns}'
 
     entrypoint = ''
-- 
2.39.5

