--- a/interface-definitions/system_console.xml.in
+++ b/interface-definitions/system_console.xml.in
@@ -12,7 +12,7 @@
             <properties>
               <help>Serial console device name</help>
               <completionHelp>
-                <script>ls -1 /dev | grep -e ttyS -e ttyAMA -e hvc</script>
+                <script>ls -1 /dev | grep -e ttyS -e ttyAMA -e hvc -e ttyFIQ</script>
                 <script>if [ -d /dev/serial/by-bus ]; then ls -1 /dev/serial/by-bus; fi</script>
               </completionHelp>
               <valueHelp>
@@ -24,6 +24,10 @@
                 <description>TTY device name, ttyAMA based</description>
               </valueHelp>
               <valueHelp>
+                <format>ttyFIQN</format>
+                <description>TTY device name, FIQ-based performance serial port</description>
+              </valueHelp>
+              <valueHelp>
                 <format>usbNbXpY</format>
                 <description>TTY device name, USB based</description>
               </valueHelp>
@@ -32,7 +36,7 @@
                 <description>Xen console</description>
               </valueHelp>
               <constraint>
-                <regex>(ttyS[0-9]+|ttyAMA[0-9]+|hvc[0-9]+|usb[0-9]+b.*)</regex>
+                <regex>(ttyS[0-9]+|ttyAMA[0-9]+|ttyFIQ[0-9]+|hvc[0-9]+|usb[0-9]+b.*)</regex>
               </constraint>
             </properties>
             <children>
@@ -46,7 +50,7 @@
                 <properties>
                   <help>Console baud rate</help>
                   <completionHelp>
-                    <list>1200 2400 4800 9600 19200 38400 57600 115200</list>
+                    <list>1200 2400 4800 9600 19200 38400 57600 115200 1500000</list>
                   </completionHelp>
                   <valueHelp>
                     <format>1200</format>
@@ -80,8 +84,12 @@
                     <format>115200</format>
                     <description>115200 bps</description>
                   </valueHelp>
+                  <valueHelp>
+                    <format>1500000</format>
+                    <description>1500000 bps</description>
+                  </valueHelp>
                   <constraint>
-                    <regex>(1200|2400|4800|9600|19200|38400|57600|115200)</regex>
+                    <regex>(1200|2400|4800|9600|19200|38400|57600|115200|1500000)</regex>
                   </constraint>
                 </properties>
                 <defaultValue>115200</defaultValue>
