Reserve first 32 MiB for NXP LS1046A firmware boundary.

diff --git a/python/vyos/system/disk.py b/python/vyos/system/disk.py
index 7ae9a15..4792350 100644
--- a/python/vyos/system/disk.py
+++ b/python/vyos/system/disk.py
@@ -62,8 +62,9 @@ def find_persistence() -> str:
 def parttable_create(drive_path: str, root_size: int) -> None:
     """Create a hybrid MBR/GPT partition table
     0-2047 first sectors are free
-    2048-4095 sectors - BIOS Boot Partition
-    4096 + 256 MB - EFI system partition
+    0-65535 (first 32 MiB) reserved for NXP LS1046A firmware boundary
+    65536-67583 sectors (32-33 MiB) - BIOS Boot Partition
+    67584 + 256 MB - EFI system partition
     Everything else till the end of a drive - Linux partition
 
     Args:
@@ -73,7 +74,7 @@ def parttable_create(drive_path: str, root_size: int) -> None:
         root_size_text: str = '+100%'
     else:
         root_size_text: str = str(root_size)
-    command = f'sgdisk -a1 -n1:2048:4095 -t1:EF02 -n2:4096:+256M -t2:EF00 \
+    command = f'sgdisk -a1 -n1:65536:67583 -t1:EF02 -n2:67584:+256M -t2:EF00 \
         -n3:0:+{root_size_text}K -t3:8300 {drive_path}'
 
     run(command)
